lib/phoenix/pubsub/redis_server.ex

```diff
@@ -0,0 +1,166 @@
+    {:reply, resp, state}
+  end
+
+  @doc """
+  Decodes binary message and fowards to local subscribers of topic
+  """
+  def handle_info({:message, _redis_topic, bin_msg, _cli_pid}, state) do
+    {_vsn, remote_node_ref, from_pid, topic, msg} = :erlang.binary_to_term(bin_msg)
+
+    if remote_node_ref == state.node_ref do
+      GenServer.call(state.local_name, {:broadcast, from_pid, topic, msg})
+    else
+      GenServer.call(state.local_name, {:broadcast, :none, topic, msg})
+    end
+
+    :eredis_sub.ack_message(state.eredis_sub_pid)
+    {:noreply, state}
+  end
+
+  def handle_info({:EXIT, _pid, {:connection_error, {:connection_error, :econnrefused}}}, state) do
+    {:noreply, state}
+  end
+
+  def handle_info({:EXIT, eredis_sub_pid, _}, %{eredis_sub_pid: eredis_sub_pid} = state) do
+    send(self, :establish_conn)
+    {:noreply, %{state | status: :disconnected, eredis_sub_pid: nil}}
+  end
+
+  @doc """
+  Connection establishment and shutdown loop
+
+  On init, an initial conection to redis is attempted when starting `:eredis_sub`
+  """
+  def handle_info(:establish_conn, state) do
+    handle_establish_conn(state)
+  end
+
+  def handle_info({:subscribed, _pattern, _client_pid}, state) do
+    :eredis_sub.ack_message(state.eredis_sub_pid)
+    {:noreply, state}
+  end
+
+  def handle_info({:eredis_connected, _client_pid}, state) do
+    Logger.info "redis connection re-established"
+    establish_success(state.eredis_sub_pid, state)
+  end
+
+  def handle_info({:eredis_disconnected, _client_pid}, state) do
+    Logger.error "lost redis connection. Attempting to reconnect..."
+    {:noreply, %{state | status: :disconnected}}
+  end
+
+  def terminate(_reason, %{status: :disconnected}) do
+    :ok
+  end
+  def terminate(_reason, state) do
+    :eredis_client.stop(state.eredis_sub_pid)
+  end
+
+  defp redis_namespace(server_name), do: "phx:#{server_name}"
+
+  defp handle_establish_conn(state) do
+    case :eredis_sub.start_link(state.opts) do
+      {:ok, eredis_sub_pid} -> establish_success(eredis_sub_pid, state)
+      _error                -> establish_failed(state)
+     end
+  end
+  defp establish_failed(state) do
+    Logger.error "unable to establish redis connection. Attempting to reconnect..."
+    :timer.send_after(@reconnect_after_ms, :establish_conn)
+    {:noreply, %{state | status: :disconnected}}
+  end
+  defp establish_success(eredis_sub_pid, state) do
+    :eredis_sub.controlling_process(eredis_sub_pid)
+    :eredis_sub.subscribe(eredis_sub_pid, [state.namespace])
+
+    {:noreply, %{state | eredis_sub_pid: eredis_sub_pid,
+                         status: :connected,
+                         node_ref: make_node_ref(state)}}
+  end
+
+  defp make_node_ref(state) do
+    :poolboy.transaction state.pool_name, fn eredis_conn ->
+      {:ok, eredis_pid} = GenServer.call(eredis_conn, :eredis)
+      {:ok, count} = :eredis.q(eredis_pid, ["INCR", "#{state.namespace}:node_counter"])
```

Comment:
Great!



*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Lead Developer*

On Thu, Feb 12, 2015 at 12:12 PM, Patrick Detlefsen <
notifications@github.com> wrote:

> In lib/phoenix/pubsub/redis_server.ex
> <https://github.com/phoenixframework/phoenix/pull/594#discussion_r24573682>
> :
>
> > +    :timer.send_after(@reconnect_after_ms, :establish_conn)
> > +    {:noreply, %{state | status: :disconnected}}
> > +  end
> > +  defp establish_success(eredis_sub_pid, state) do
> > +    :eredis_sub.controlling_process(eredis_sub_pid)
> > +    :eredis_sub.subscribe(eredis_sub_pid, [state.namespace])
> > +
> > +    {:noreply, %{state | eredis_sub_pid: eredis_sub_pid,
> > +                         status: :connected,
> > +                         node_ref: make_node_ref(state)}}
> > +  end
> > +
> > +  defp make_node_ref(state) do
> > +    :poolboy.transaction state.pool_name, fn eredis_conn ->
> > +      {:ok, eredis_pid} = GenServer.call(eredis_conn, :eredis)
> > +      {:ok, count} = :eredis.q(eredis_pid, ["INCR", "#{state.namespace}:node_counter"])
>
> it's a 64-bit signed integer so it should take a long time to overflow
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/phoenixframework/phoenix/pull/594/files#r24573682>.
>